### PR TITLE
[move-ide] Modified move-analyzer installation procedure to utilize versioning

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This goal of this PR is to allow friction-less updates of the Move analyzer binary, particularly for developers that utilize Sui CLI, which should be the majority. We still want friction-less Move analyzer binary installation for developers that want to only use the IDE and do not have a local CLI installation.

Move analyzer can now be installed in three different way:
1. it's bundled with the VSCode extension and if no local Move analyzer installation exists (or local installations have a lower version than the bundled one), it gets installed as part of VSCode Move extension activation process
2. it's bundled with Sui's CLI binary (if this binary has the highest version of all those installed, it will be used)
3. it can be built locally (or downloaded from GitHub) and placed in the default location (`~/.sui/bin`) or in a location specified by the user via VSCode extension's option

In cases 1 and 3 what gets ultimately used by the VSCode extension is the "standalone" Move analyzer binary (`move-analyzer` binary containing just the LSP server implementation). In case 2 what gets ultimately used is the Sui CLI which contains Move analyzer binary (and which is invoked via a special `analyzer` Sui CLI option).

Once this PR lands, the following will be enabled:
- if the first and only thing a developer does is to install VSCode extension from the marketplace, it's ready to use with all the functionality available
- if the developer installed Sui CLI, with each update of the CLI binary a new version of Move analyzer will be available even if the extension in the marketplace is not updated (which helps keeping parity between compiler version in Move analyzer binary and compiler version in the CLI binary, to avoid discrepancy with respect to features, error messages, etc.)

## Test plan 

Tested manually the following scenarios:

1. Check cases when no higher installed version exists:
- install bundled extension when no local installation exists
- use locally installed standalone version (default path) if bundled has the same version (no Sui version installed)
- use locally installed standalone version (user-specified path) if bundled has the same version (no Sui version installed)
- use locally installed Sui version (on system path) if bundled has the same version (no standalone version installed)
- use locally installed standalone version (user-specified path) if bundled has the same version and installed Sui one has the same version
- use locally installed Sui version (on user-specified path) if bundled has the same version (no standalone version installed)
- report error if no binary at user-specified path (no other local installs)
- report error if no binary at user-specified path (Sui version installed on system path)
2. Check cases when a higher installed version exists:
- use locally installed standalone version (default path) if bundled has lower version (no Sui version installed)
- use locally installed Sui version (on system path) if bundled has lower version (no standalone version installed)
3. Check cases when a lower installed version exists:
- install bundled version if installed standalone version is lower (no Sui version installed)
- install bundled version if installed Sui version is lower (no standalone version installed)
- install bundled version if both installed Sui version and standalone version are lower

### CODE REVIEW

File: external-crates/move/crates/move-analyzer/editors/code/package-lock.json [hunk 1]
```diff
@@ -1,3954 +1,3936 @@
 {
-	"name": "move",
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/editors/code/package-lock.json:2]: Largely unrelated (did not re-generate this file after re-formatting `package.json` in https://github.com/MystenLabs/sui/pull/17863)

File: external-crates/move/crates/move-analyzer/editors/code/src/configuration.ts [hunk 2]
```diff
@@ -65,98 +69,12 @@ export class Configuration {
         return path.resolve(suiPath);
     }
 
-    /**
-     * Installs language server binary in the default location unless a user-specified
-     * (but not default) server location already contains a server binary.
-     *
-     * @returns `true` if server binary installation succeeded, `false` otherwise.
-     */
-    async installServerBinary(extensionContext: vscode.ExtensionContext): Promise<boolean> {
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/editors/code/src/configuration.ts:74]: Moved to `context.ts`

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts [hunk 3]
```diff
@@ -177,25 +203,199 @@ export class Context {
...
         });
 
     }
+
+    /**
+     * Installs language server binary in the default location if needed.
+     * On a happy path we just compare versions of installed and bundled
+     * binaries and don't do any actual file operations at all.
+     *
+     * The actual algorithm deciding on how this works as follows.
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/editors/code/src/context.ts:243]: There are many ways to slice it and I think this one makes sense but I am open to other suggestions

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts [hunk 4]
```diff
@@ -177,25 +203,199 @@ export class Context {
...
+        // What's the highest version installed? Also track path and arguments to run analyzer
+        // with the highest version
+        let highestVersion = null;
+        if (standaloneVersion !== null) {
+            highestVersion = standaloneVersion;
+            this.resolvedServerPath = this.configuration.serverPath;
+            this.resolvedServerArgs = serverArgs;
+        }
+        if (cliVersion !== null) {
+            if (standaloneVersion === null || semver.gt(cliVersion, standaloneVersion)) {
```
[the remark this answers is not in the document]
Comment (amnn) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts:316]: Continuing this thought, it's possible that we can simplify the logic a bit more by treating it all as a cascade of tests:

```ts
let highestVersion = null;
if (standaloneVersion !== null && (highestVersion === null || semver.gt(standaloneVersion, highestVersion))) {
   /* ... standalone is highest now ... */ 
}

if (cliVersion !== null && (highestVersion === null || semver.gt(cliVersion, highestVersion))) {
    /* ... CLI version is highest now ... */
}

if (bundledVersion !== null && (highestVersion === null || semver.gt(bundledVersion, highestVersion))) {
    /* ... bundled version is highest now, and also install it ... */
}

if (highestVersion === null) {
    /* Error about pre-built not being available */
}
```

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts [hunk 5]
```diff
@@ -2,22 +2,55 @@
...
+import { assert } from 'console';
 import { IndentAction } from 'vscode';
 
+function semanticVersion(path: string, args?: readonly string[]): semver.SemVer | null {
+    const versionString = childProcess.spawnSync(
+        path, args, { encoding: 'utf8' },
+    );
+    if (versionString.stdout) {
+        // Version string looks as follows: 'COMMAND_NAME SEMVER-SHA'
+        const versionStringWords = versionString.stdout.split(' ');
```
Comment (amnn) [inline external-crates/move/crates/move-analyzer/editors/code/src/context.ts:23]: You may want to set the split limit as well, to make sure you get at most two parts?

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (amnn) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts:316]: Also installation probably belongs in its own helper function.

**Review by amnn — COMMENTED** [review]
Left a suggestion to simplify the cascading logic for picking versions, I think the current version is correct, but it's a little hard to follow with all the different combinations of conditions floating around -- let me know what you think about the simplification.

File: external-crates/move/crates/move-analyzer/editors/code/src/main.ts [hunk 6]
```diff
@@ -117,64 +117,14 @@ async function testProject(context: Readonly<Context>): Promise<void> {
...
-    // An error here -- for example, if the path to the `move-analyzer` binary that the user
-    // specified in their settings is not valid -- prevents the extension from providing any
-    // more utility, so return early.
-    if (context instanceof Error) {
-        void vscode.window.showErrorMessage(
-            `Could not activate Move: ${context.message}.`,
-        );
+    const context = new Context(extensionContext);
+    const success = await context.installServerBinary(extensionContext);
+    if (!success) {
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/editors/code/src/main.ts:126]: Makes sense. It's signaled to the user via info window pop-ups triggered inside `installServerBinary` but logging info message as well is certainly not going to hurt

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts]: > Continuing this thought, it's possible that we can simplify the logic a bit more by treating it all as a cascade of tests

Makes a lot of sense! Done!

**Comment by awelc** [comment]
> Left a suggestion to simplify the cascading logic for picking versions, I think the current version is correct, but it's a little hard to follow with all the different combinations of conditions floating around -- let me know what you think about the simplification.

Thank you for the suggestion which I gladly followed!

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts [hunk 7]
```diff
@@ -177,25 +203,176 @@ export class Context {
...
+                } catch (err) {
+                    // Handle error
+                    log.info(String(err));
                 }
             }
         });
 
     }
+
+    async installBundledBinary(bundledServerPath: vscode.Uri): Promise<void> {
```
Comment (stefan-mysten) [inline external-crates/move/crates/move-analyzer/editors/code/src/context.ts:238]: Would some logs make sense here? For example, removing the existing binary might be useful... 

On another issue, this would fail if there is no disk space, right?

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts [hunk 8]
```diff
@@ -177,25 +203,176 @@ export class Context {
...
+            // User has overwritten default server path (need to compare paths as comparing URIs fails
+            // for some reason even if one is initialized from another).
+            if (standaloneVersion === null) {
+                // The server binary on user-overwritten path does not return version number.
+                // Need to use modal messages, otherwise the promise returned from is not resolved
+                // (and the extension blocks), which can confusing.
+                const items: vscode.MessageItem = { title: 'OK', isCloseAffordance: true };
+                await vscode.window.showInformationMessage(
+                    `The move-analyzer binary at the user-specified path ('${this.configuration.serverPath}') ` +
+                    'is not working. Put the corrent binary in the path or reset user settings and ' +
```
Comment (stefan-mysten) [inline external-crates/move/crates/move-analyzer/editors/code/src/context.ts:317]:
```suggestion
                    'is not working. Put the correct binary in the path or reset user settings and ' +
```

Comment (stefan-mysten) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts:317]: Which is the right binary and which is the right path? Can the error message provide some direction there?

**Review by stefan-mysten — APPROVED** [review]

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts (hunk 8 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts]: > Which is the right binary and which is the right path? Can the error message provide some direction there?

Good point! I followed a similar path and directed the user to consult extension's README file for troubleshooting

File: external-crates/move/crates/move-analyzer/editors/code/src/context.ts (hunk 7 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/editors/code/src/context.ts:238]: > Would some logs make sense here? For example, removing the existing binary might be useful...

Added additional info to the logs

> On another issue, this would fail if there is no disk space, right?

Indeed, but presumably all sorts of things will start failing on the system at this point so I'd think we don't have to worry about handling this case explicitly here
